File: algorithms/math/mobius-function/scala/MobiusFunction.scala
object MobiusFunction {
  def mobiusFunction(n: Int): Int = {
    val mu = Array.fill(n + 1)(1)
    mu(1) = 1
    val isPrime = Array.fill(n + 1)(true)
    if (n >= 0) {
      isPrime(0) = false
    }
    if (n >= 1) {
      isPrime(1) = false
    }

    for (i <- 2 to n) {
      if (isPrime(i)) {
        var j = i
        while (j <= n) {
          if (j != i) {
            isPrime(j) = false
          }
          mu(j) = -mu(j)
          j += i
        }
        val i2 = i.toLong * i
        var k = i2
        while (k <= n) {
          mu(k.toInt) = 0
          k += i2
        }
      }
    }
    mu.drop(1).sum
  }

  def main(args: Array[String]): Unit = {
    println(mobiusFunction(1))
    println(mobiusFunction(10))
    println(mobiusFunction(50))
  }
}

File: algorithms/math/mobius-function/rust/mobius_function.rs
fn mobius_function(n: usize) -> i32 {
    if n == 0 {
        return 0;
    }

    let mut mu = vec![1i32; n + 1];
    let mut is_prime = vec![true; n + 1];
    mu[0] = 0;

    for i in 2..=n {
        if is_prime[i] {
            let mut j = i;
            while j <= n {
                is_prime[j] = false;
                mu[j] = -mu[j];
                j += i;
            }
            let i2 = i * i;
            let mut k = i2;
            while k <= n {
                mu[k] = 0;
                k += i2;
            }
        }
    }

    mu[1..].iter().sum()
}

fn main() {
    println!("{}", mobius_function(1));
    println!("{}", mobius_function(10));
    println!("{}", mobius_function(50));
}
